### PR TITLE
[AMORO-1899] add init schema in mysql

### LINKED ISSUE #1899: [Improvement]: add mysql schema init
### Search before asking

- [X] I have searched in the [issues](https://github.com/NetEase/amoro/issues?q=is%3Aissue) and found no similar issues.

### What would you like to be improved?

auto init ams schema in mysql.

### How should we improve?

_No response_

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Subtasks

_No response_

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by Kyofin**
## Why are the changes needed?

Close #1899 

## Brief change log

- add init mysql schema srcipt 

## How was this patch tested?

- [ ] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [x] Run test locally before making a pull request

## Documentation

- Does this pull request introduce a new feature? (yes / no)
- If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented)

### CODE REVIEW

**Review by baiyangtx — APPROVED** [review]
LGTM.
Thank you for submitting your Pull Request, your contribution is greatly appreciated!

File: ams/server/src/main/java/com/netease/arctic/server/persistence/SqlSessionFactoryProvider.java [hunk 1]
```diff
@@ -136,11 +158,17 @@ private String getDerbyInitSqlScriptPath() {
     }
     return scriptUrl.getPath();
   }
-
+  private String getMysqlInitSqlScriptPath() {
```
Comment (shidayang) [inline ams/server/src/main/java/com/netease/arctic/server/persistence/SqlSessionFactoryProvider.java:161]: Could you reuse getDerbyInitSqlScriptPath?

**Comment by wangtaohz** [comment]
I found that there are some check style issues. 

You can run `mvn checkstyle:check` locally to check the result. 😊 @Kyofin

File: ams/server/src/main/java/com/netease/arctic/server/persistence/SqlSessionFactoryProvider.java [hunk 2]
```diff
@@ -84,7 +86,7 @@ public void init(Configurations config) {
     dataSource.setNumTestsPerEvictionRun(BaseObjectPoolConfig.DEFAULT_NUM_TESTS_PER_EVICTION_RUN);
     dataSource.setTestOnReturn(BaseObjectPoolConfig.DEFAULT_TEST_ON_RETURN);
     dataSource.setSoftMinEvictableIdleTimeMillis(
-        BaseObjectPoolConfig.DEFAULT_SOFT_MIN_EVICTABLE_IDLE_TIME.toMillis());
+            BaseObjectPoolConfig.DEFAULT_SOFT_MIN_EVICTABLE_IDLE_TIME.toMillis());
```
Comment (zhoujinsong) [inline ams/server/src/main/java/com/netease/arctic/server/persistence/SqlSessionFactoryProvider.java:89]: @Kyofin Amoro project uses 2 spaces for indent and tab and 4 spaces for continuation indent for java codes.

Could you please change the configuration and reformat the codes?

File: ams/server/src/main/java/com/netease/arctic/server/persistence/SqlSessionFactoryProvider.java [hunk 3]
```diff
@@ -109,38 +111,73 @@ public void init(Configurations config) {
 
   // create tables for derby database type
   private void createTablesIfNeed(Configurations config) {
-    if (ArcticManagementConf.DB_TYPE_DERBY.equals(config.getString(ArcticManagementConf.DB_TYPE))) {
+    String dbTypeConfig = config.getString(ArcticManagementConf.DB_TYPE);
```
Comment (zhoujinsong) [inline ams/server/src/main/java/com/netease/arctic/server/persistence/SqlSessionFactoryProvider.java:114]: The codes here seem to have significant duplication, with only the SQL for checking if tables have already been initialized is different for different types of databases. The rest of the logic appears to be the same.

Can you see if you can refactor the code to reduce code duplication?

File: ams/server/src/main/java/com/netease/arctic/server/persistence/SqlSessionFactoryProvider.java [hunk 4]
```diff
@@ -109,38 +111,73 @@ public void init(Configurations config) {
...
+      }
     }
   }
 
-  private String getDerbyInitSqlScriptPath() {
-    URL scriptUrl = ClassLoader.getSystemResource(DERBY_INIT_SQL_SCRIPT);
-    if (scriptUrl == null) {
-      throw new IllegalStateException("Cannot find derby init sql script:" + DERBY_INIT_SQL_SCRIPT);
+  private String getInitSqlScriptPath(String type) {
+    if (type.equals(ArcticManagementConf.DB_TYPE_MYSQL)) {
```
Comment (zhoujinsong) [inline ams/server/src/main/java/com/netease/arctic/server/persistence/SqlSessionFactoryProvider.java:159]: There seems to be an issue with repeated codes here too.

File: ams/server/src/main/java/com/netease/arctic/server/persistence/SqlSessionFactoryProvider.java [hunk 5]
```diff
@@ -107,40 +109,58 @@ public void init(Configurations config) {
...
+    }
+    URL scriptUrl = ClassLoader.getSystemResource(scriptPath);
     if (scriptUrl == null) {
-      throw new IllegalStateException("Cannot find derby init sql script:" + DERBY_INIT_SQL_SCRIPT);
+      throw new IllegalStateException("Cannot find  init sql script:" + scriptPath);
     }
     return scriptUrl.getPath();
   }
 
+
```
Comment (zhoujinsong) [inline ams/server/src/main/java/com/netease/arctic/server/persistence/SqlSessionFactoryProvider.java:158]: Can you remove the duplicate empty lines here?

**Review by zhoujinsong — APPROVED** [review]
LGTM.

File: ams/server/src/main/java/com/netease/arctic/server/persistence/SqlSessionFactoryProvider.java [hunk 6]
```diff
@@ -107,32 +109,47 @@ public void init(Configurations config) {
...
-          }
-        }
-      } catch (Exception e) {
-        throw new IllegalStateException("Create derby tables failed", e);
+    String dbTypeConfig = config.getString(ArcticManagementConf.DB_TYPE);
+    String query = "";
+
+    try (SqlSession sqlSession = get().openSession(true);
+         Connection connection = sqlSession.getConnection();
+         Statement statement = connection.createStatement();) {
```
Comment (wangtaohz) [inline ams/server/src/main/java/com/netease/arctic/server/persistence/SqlSessionFactoryProvider.java:123]: The `;` is redundant.

File: ams/server/src/main/java/com/netease/arctic/server/persistence/SqlSessionFactoryProvider.java [hunk 7]
```diff
@@ -107,32 +109,47 @@ public void init(Configurations config) {
...
+      if (ArcticManagementConf.DB_TYPE_DERBY.equals(dbTypeConfig)) {
+        query = "SELECT 1 FROM SYS.SYSTABLES WHERE TABLENAME = 'CATALOG_METADATA'";
+      } else if (ArcticManagementConf.DB_TYPE_MYSQL.equals(dbTypeConfig)) {
+        query = String.format(
+            "SELECT COUNT(*) FROM information_schema.tables WHERE table_schema = '%s' AND table_name = '%s'",
```
Comment (wangtaohz) [inline ams/server/src/main/java/com/netease/arctic/server/persistence/SqlSessionFactoryProvider.java:128]: I tested locally with MySQL, but it didn't work.

I think we should change the `SELECT COUNT(*) FROM` to `SELECT * FROM`. @Kyofin

**Review by shidayang — APPROVED** [review]

File: ams/server/src/main/java/com/netease/arctic/server/persistence/SqlSessionFactoryProvider.java (hunk 7 above)
Comment (wangtaohz) [reply ams/server/src/main/java/com/netease/arctic/server/persistence/SqlSessionFactoryProvider.java:128]: Maybe `SELECT 1` is better, keepping the same as Derby.

File: ams/server/src/main/java/com/netease/arctic/server/persistence/SqlSessionFactoryProvider.java [hunk 8]
```diff
@@ -107,32 +109,47 @@ public void init(Configurations config) {
...
+            connection.getCatalog(), "CATALOG_METADATA");
+      }
+      ResultSet rs = statement.executeQuery(query);
```
Comment (wangtaohz) [inline ams/server/src/main/java/com/netease/arctic/server/persistence/SqlSessionFactoryProvider.java:131]: The `ResultSet` may also need `close`.

**Review by wangtaohz — COMMENTED** [review]
As this PR has already been merged into master, we can create a new PR to address the remaining issues.

Also, the docs could be simplified by removing the section on initializing the tables of MySQL .
https://github.com/NetEase/amoro/blob/master/docs/admin-guides/deployment.md?plain=1#L134

image

**Pull request merged**